pkg/csi/service/common/commonco/k8sorchestrator/k8sorchestrator.go

```diff
@@ -236,14 +225,108 @@ func initFSS(ctx context.Context, k8sClient clientset.Interface, controllerClust
+			} else {
+				doesSvFssCRExist = true
+				// Store SV FSS values in cache
+				for _, svFSS := range svFssCR.Spec.FeatureStates {
+					k8sOrchestratorInstance.supervisorFSS.featureStates[svFSS.Name] = strconv.FormatBool(svFSS.Enabled)
+				}
+				log.Infof("New supervisor feature states values stored successfully from %s CR object: %v",
+					featurestates.SVFeatureStateCRName, k8sOrchestratorInstance.supervisorFSS.featureStates)
+			}
+			// Set up namespaced listener for cnscsisvfeaturestate CR even if it doesn't exist
+			// This is required in cases where TKG is on a newer build and SV is being upgraded from an older version
+			// TODO: check if this will break backward compatibility as this CR is not yet registered in an old SV
+			err = k8sOrchestratorInstance.informerManager.AddFSSListener(ctx,
+				featurestates.CRDGroupName,
+				internalapis.Version,
+				featurestates.CRDName,
+				svNamespace,
+				// Add
+				func(obj interface{}) {
+					fssCRAdded(obj)
+				}, // Update
+				func(oldObj interface{}, newObj interface{}) {
+					fssCRAdded(newObj)
+				}, // Delete
+				func(obj interface{}) {
+					fssCRDeleted(obj)
+				})
+			if err != nil {
+				log.Errorf("Failed to listen on cnscsisvfeaturestate CR. Error - %+v", err)
+				return err
+			}
+			log.Infof("Informer to watch on %s CR started", featurestates.CRDSingular)
+		}
+	}
+	// Initialize supervisor FSS map values using configmap in Supervisor cluster flavor
+	// or in guest cluster flavor only if cnscsisvfeaturestate CR is not found
+	if controllerClusterFlavor == cnstypes.CnsClusterFlavorWorkload || (!doesSvFssCRExist && controllerClusterFlavor == cnstypes.CnsClusterFlavorGuest) {
+		if k8sOrchestratorInstance.supervisorFSS.configMapName != "" && k8sOrchestratorInstance.supervisorFSS.configMapNamespace != "" {
+			// Retrieve configmap
+			fssConfigMap, err = k8sClient.CoreV1().ConfigMaps(k8sOrchestratorInstance.supervisorFSS.configMapNamespace).Get(
+				ctx, k8sOrchestratorInstance.supervisorFSS.configMapName, metav1.GetOptions{})
+			if err != nil {
+				log.Errorf("failed to fetch configmap %s from namespace %s. Setting the supervisor feature states to default values: %v. Error: %v",
+					k8sOrchestratorInstance.supervisorFSS.configMapName, k8sOrchestratorInstance.supervisorFSS.configMapNamespace,
+					k8sOrchestratorInstance.supervisorFSS, err)
+			} else {
+				// Update values
+				k8sOrchestratorInstance.supervisorFSS.featureStates = fssConfigMap.Data
+				log.Infof("New supervisor feature states values stored successfully: %v", k8sOrchestratorInstance.supervisorFSS.featureStates)
+			}
+		}
+	}
+	// Set up kubernetes configmap listener for CSI namespace
 	k8sOrchestratorInstance.informerManager.AddConfigMapListener(ctx, k8sClient, configMapNamespaceToListen,
 		// Add
 		func(obj interface{}) {
 			configMapAdded(obj)
 		}, // Update
 		func(oldObj interface{}, newObj interface{}) {
-			configMapUpdated(oldObj, newObj)
+			configMapAdded(newObj)
```

Comment:
Can we keep `configMapUpdated(oldObj, newObj)`, and create a new func to move common code for `configMapAdded` and `configMapUpdated`?

Reply:
Done